py3status/modules/wwan_status_nm.py

```diff
@@ -0,0 +1,133 @@
+        use 'busctl introspect org.freedesktop.ModemManager1 \
+            /org/freedesktop/ModemManager1/Modem/0'
+        and read '.EquipmentIdentifier')
+        (default None)
+
+Color options:
+    color_bad: Error or no connection
+    color_degraded: Low generation connection eg 2G
+    color_good: Good connection
+
+Requires:
+    ModemManager
+    NetworkManager
+    pydbus
+
+@author Cyril Levis <levis.cyril@gmail.com>
+
+SAMPLE OUTPUT
+{'color': '#00FF00', 'full_text': u'Bouygues Telecom 4G (19%)'}
+
+off
+{'color': '#FF0000', 'full_text': u'Bouygues Telecom 4G (12%)'}
+"""
+
+from pydbus import SystemBus
+
+STRING_WRONG_MODEM = "wrong or any modem"
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    cache_timeout = 5
+    consider_3G_degraded = False
+    format_down = 'WWAN: {operator} {netgen} ({signal}%)'
+    format_error = 'WWAN: {error}'
+    format_up = 'WWAN: {operator} {netgen} ({signal}%)'
+    modem = None
+
+    def wwan_status_nm(self):
+        response = {}
+        response['cached_until'] = self.py3.time_in(self.cache_timeout)
+
+        bus = SystemBus()
+        for id in range(0, 20):
+            """
+            find the modem
+            """
+            device = 'Modem/' + str(id)
+
+            try:
+                proxy = bus.get('.ModemManager1', device)
+
+                eqid = str(proxy.EquipmentIdentifier)
+
+                if (self.modem is not None
+                        and eqid == self.modem) or (self.modem is None):
+
+                    data = {
+                        'state': proxy.State,
+                        'signal': str(proxy.SignalQuality[0]),
+                        'modes': proxy.CurrentModes[0],
+                        'operator': proxy.OperatorName
+                    }
+
+                    netgen = self._get_capabilities(data['modes'])
+
+                    data['netgen'] = netgen[1]
+
+                    if data['state'] == 11:
+                        response['full_text'] = self.py3.safe_format(
+                            self.format_up, data)
+                        """
+                        green color if 4G, else yellow
+                        """
+                        if netgen[0] == 4:
+                            response['color'] = self.py3.COLOR_GOOD
+                        elif netgen[0] == 3 and self.consider_3G_degraded is False:
+                            response['color'] = self.py3.COLOR_GOOD
+                        else:
+                            response['color'] = self.py3.COLOR_DEGRADED
+
+                    else:
+                        response['full_text'] = self.py3.safe_format(
+                            self.format_down, data)
+                        response['color'] = self.py3.COLOR_BAD
+
+                    return response
+
+                else:
+                    self.py3.error(STRING_WRONG_MODEM)
+
+            except:
+                pass
```

Comment:
This is likely the culprit for `raise TypeError('response should be a dict')`.

Reply:
i must send an empty string so?

Reply:
Sorry. Yes. We need to return at least this.
```

diff --git a/py3status/modules/wwan_status_nm.py b/py3status/modules/wwan_status_nm.py
index a2c3f656..a211c120 100644
--- a/py3status/modules/wwan_status_nm.py
+++ b/py3status/modules/wwan_status_nm.py
@@ -114,7 +114,8 @@ class Py3status:
                     self.py3.error(STRING_WRONG_MODEM)

             except:
-                pass
+                response['full_text'] = ''
+                return response

     def _get_capabilities(self, c):
         """
```
It would be more readable to return a dict outside of everything. Use `break` to get out of the loop too.